### PR TITLE
Support mixins

### LINKED ISSUE #63: Add documentation on how to use with prop types defined in multiple components/mixins
I have a component `foo`, and a mixin `foo-common` that adds a common API to that component. Both use ember-prop-types.

The component `foo` has a `propTypes` attribute and a `getDefaultProps()` method. So does the mixin `foo-common`.

There are two problems with the result:

1. The `propTypes` for the `foo-common` are not set, because `ember-prop-types` does not define propTypes as a `concatenatedProperty`. The `propTypes` for `foo` override those in `foo-common`.
2. The default property values for the mixin are not set, because `getDefaultProps` from `foo` similarly overrides `foo-common`.

PR pending.

### PR DESCRIPTION
**Pull request by justinbowes**
**This project uses [semver](http://semver.org), please check the scope of this pr:**

- [ ] #patch# - backwards-compatible bug fix
- [ ] #minor# - adding functionality in a backwards-compatible manner
- [x] #major# - incompatible API change

Resolves #63 

# CHANGELOG

* Support cases where components and mixins both use this mixin.

### CODE REVIEW

**Comment by justinbowes** [comment]
Added tests for getDefaultProps as well, which were lacking.

File: addon/mixins/prop-types.js [hunk 1]
```diff
@@ -33,41 +33,47 @@ const helpers = {
...
-    if (!propTypes) {
-      return
-    }
-
-    Object.keys(propTypes).forEach(name => {
-      const def = propTypes[name]
-
-      if (def === undefined) {
-        Ember.Logger.warn(`propType for ${name} is unknown`)
+    const propTypes = [].concat(ctx.get('propTypes'))
```
Comment (sandersky) [inline addon/mixins/prop-types.js]: Why this change instead of just returning when it isn't present?

Comment (sandersky) [inline addon/mixins/prop-types.js]: Never mind I guess this has to do with the concatenation.

Comment (sandersky) [inline addon/mixins/prop-types.js]: I'm thinking we should actually name this variable `propTypesArray` and change `property` to `propTypes`. When I see this variable called `propTypes` I think it is of the format:

```js
{
  foo: PropTypes.string,
  bar: PropTypes.number
}
```

when in actuality if I'm understanding this change correctly is of the format:

```js
[
  {
    foo: PropTypes.string
  },
  {
    bar: PropTypes.number
  }
]
```

File: addon/mixins/prop-types.js [hunk 2]
```diff
@@ -33,41 +33,47 @@ const helpers = {
...
-      return
-    }
-
-    Object.keys(propTypes).forEach(name => {
-      const def = propTypes[name]
-
-      if (def === undefined) {
-        Ember.Logger.warn(`propType for ${name} is unknown`)
+    const propTypesArray = [].concat(ctx.get('propTypes'))
+    propTypesArray.forEach((property) => {
```
Comment (sandersky) [inline addon/mixins/prop-types.js]: Still would like to see `property` renamed to `propTypes`

Comment (justinbowes) [inline addon/mixins/prop-types.js]: doh, missed that

File: addon/mixins/prop-types.js [hunk 3]
```diff
@@ -33,41 +33,47 @@ const helpers = {
...
+        }
+
+        helpers.validateProperty(ctx, name, def)
+      })
     })
   }
 }
 
 export default Ember.Mixin.create({
+  concatenatedProperties: ['propTypes', 'getDefaultProps'],
```
Comment (sglanzer) [inline addon/mixins/prop-types.js]: Love it, really nice

**Comment by sglanzer** [comment]
Looks great - I assume the only downstream implication will be that context where we used to call _super in a getDefaultProps, we just stop doing that...

It wasn't part of the "official" API for how to use this anyway, but any idea what will happen in those cases?

**Comment by sandersky** [comment]
@sglanzer that is a valid concern, we should figure that out before merging. If it breaks that then this needs to be a major.

**Comment by sglanzer** [comment]
`ember-frost-modals` uses that pattern, we could run this fork against it and find out, the demo is pretty complete

**Comment by sglanzer** [comment]
My _guess_ is that it'll just result in inefficiency (concat/overwrite multiple times), but no functional impact

**Comment by justinbowes** [comment]
My concern is that somebody is using `this._super` in `getDefaultProps` to delete a default value -- this is no longer possible. 

Then again, in the hypothetical case where you needed to remove a key from defaultProps, you'd probably want to remove a `propType` key as well, and I didn't see a good way to do that either before or now (corrections welcomed).

**Comment by sglanzer** [comment]
When anyone has time we can confirm this fairly quickly using the modal repo - I just don't have time currently :)

**Comment by justinbowes** [comment]
`ember-frost-modals` makes an assumption that I broke with this PR:

https://github.com/ciena-frost/ember-frost-modal/blob/master/addon/components/dialogs/about.js#L38

```

  getDefaultProps () {
    let defaultProps = this._super()

    assign(defaultProps, {
      animation: about,
      classModifier: 'about',
      closeOnOutsideClick: true,
      modal: 'frost-modal-about-dialog'
    })

    return defaultProps
  },
```

One option is to change the way `ember-frost-modals` works, which is relatively straightforward.

Another way did occur to me. I could implement a "base" `getDefaultProps()` in `ember-prop-types` and have it return `{}`:
* [+] Maintains API entirely and avoids major version bump. 
* [+] Simplifies default props processing a tiny bit -- we know there will always be a `getDefaultProps` array in this case.
* [-] Adds an extra function call to every initialization

Thoughts?

**Comment by sglanzer** [comment]
Thanks for checking on that @justinbowes - I'm 👍 for avoiding a major, if the perf is a concern then we could do this change in the 2.x branch with a deprecation notice and merge your original in a 3.x branch

I'm ok with just going ahead with the `base` and merging into 2.x for now, then worrying about a perf improvement later

**Comment by sglanzer** [comment]
I'd say "just do it, how many people are going to be impacted" - but that's burned me too many times this month and I don't want to be evil to people downstream :)

**Comment by sglanzer** [comment]
LGTM

**Comment by sglanzer** [comment]
Approved - sigh, tried to add flair to my message, pullapprove doesn't recognize it...sad panda

**Comment by justinbowes** [comment]
Please hold off on this -- it's a clean pass here but after pulling the latest commit into `modals` I'm still not seeing the expected results.

**Comment by justinbowes** [comment]
The changes do maintain compatibility on inheritance cases but not in mixins (`this._super` on a component doesn't call a mixin method; it calls into the extended class).

I have edited the template to change this to a major. If we want to hold off on this due to LTS, you can emulate this approach using your own concatenated properties (e.g. `propTypesContribution`, `getDefaultPropsContribution`).

**Comment by sglanzer** [comment]
I think we should merge this and just hold off on immediate upgrades to it until we do a sweep for the LTS - @sandersky?

**Comment by sglanzer** [comment]
Spoke with Matt, he's on board with the plan
